staging/src/k8s.io/client-go/informers/factory.go

```diff
@@ -195,6 +197,81 @@ func (f *sharedInformerFactory) WaitForCacheSync(stopCh <-chan struct{}) map[ref
 	return res
 }
 
+func (f *sharedInformerFactory) WaitForCacheSyncWithContext(ctx context.Context) CacheSyncResult {
+	informers := func() map[reflect.Type]cache.SharedIndexInformer {
+		f.lock.Lock()
+		defer f.lock.Unlock()
+
+		informers := map[reflect.Type]cache.SharedIndexInformer{}
+		for informerType, informer := range f.informers {
+			if f.startedInformers[informerType] {
+				informers[informerType] = informer
+			}
+		}
+		return informers
+	}()
+
+	// Wait for informers to sync, without polling.
+	cacheSyncs := make([]cache.NamedSyncer, 0, len(informers))
+	for _, informer := range informers {
+		cacheSyncs = append(cacheSyncs, informer.NamedHasSynced())
+	}
+	cache.WaitFor(ctx, "" /* no logging */, cacheSyncs...)
+
+	res := CacheSyncResult{
+		Synced: make(map[reflect.Type]bool, len(informers)),
+	}
+	failed := false
+	for informType, informer := range informers {
+		hasSynced := informer.HasSynced()
+		if !hasSynced {
+			failed = true
+		}
+		res.Synced[informType] = hasSynced
+	}
+	if failed {
+		// context.Cause is more informative than ctx.Err().
+		// This must be non-nil, otherwise WaitFor wouldn't have stopped
+		// prematurely.
+		res.Err = context.Cause(ctx)
+	}
+
+	return res
+}
+
+// CacheSyncResult is the result of WaitForCacheSyncWithContext.
+type CacheSyncResult struct {
```

Comment:
I probably should move this to an apimachinery package. Then I can also add a unit test (missing right now).

Reply:
SG.  I was also noticing the duplicate implementations of `WaitForCacheSyncWithContext` here and in apiextensions.  Would be good to do something about that at some point.  I'm fine with a tracking issue for now.